#3343586565e3abd817ac131ab4a328dace71c5a7
diff --git a/include/camera/camera.h b/include/camera/camera.h
index 2d1312b..70cbc1a 100755
--- a/include/camera/camera.h
+++ b/include/camera/camera.h
@@ -9,10 +9,13 @@ enum{
 };
 
 typedef enum {
-	ImageQuality1M = 0,
-	ImageQuality3M,
-	ImageQuality5M,
-	ImageQuality8M,
+	ImageQualityQQVGA = 0, /*160---120*/
+	ImageQualityQVGA,	   /*320---240*/
+	ImageQualityVGA,	   /*640---480*/
+	ImageQuality1M,		   /*1280--720*/
+	ImageQuality3M,		   /*1920-1080*/
+	ImageQuality5M,	       /*2560-1920*/
+	ImageQuality8M,	       /*3264-2448*/
 }ImageQuality_t;
 
 
diff --git a/include/storage/StorageManager.h b/include/storage/StorageManager.h
index dcb05b0..d2aa225 100755
--- a/include/storage/StorageManager.h
+++ b/include/storage/StorageManager.h
@@ -132,6 +132,7 @@ public:
 	void dbAddFileByTransaction(Elem *elem);
 	void dbDelFile(char const *file);
 	int generateFile(time_t &now, String8 &file, int cam_type, fileType_t type);
+	int generateFile2(time_t now, String8 &file, int cam_type, fileType_t type);
 	int saveVideoFrameToBmpFile(VideoFrame* videoFrame, const char* bmpFile);
 	int savePicture(void* data, int size, int fd);
 	int checkDirs();
diff --git a/include/window/windows.h b/include/window/windows.h
index b10d18b..fa3b843 100755
--- a/include/window/windows.h
+++ b/include/window/windows.h
@@ -149,6 +149,7 @@ class RecordPreview : public MainWindow
 {
 private:
 	bool SdcardState;
+	bool isThumb;
 public:
 	RecordPreview(CdrMain* cdrMain);
 	~RecordPreview();
diff --git a/src/camera/CdrCamera.cpp b/src/camera/CdrCamera.cpp
index 23a838a..1451bae 100755
--- a/src/camera/CdrCamera.cpp
+++ b/src/camera/CdrCamera.cpp
@@ -286,6 +286,24 @@ void CdrCamera::setImageQuality(ImageQuality_t quality)
 	//params.dump();
 
 	switch(quality) {
+	case ImageQualityQQVGA:
+		{
+			w = 160;
+			h = 120;
+		}
+		break;
+	case ImageQualityQVGA:
+		{
+			w = 320;
+			h = 240;
+		}
+		break;
+	case ImageQualityVGA:
+		{
+			w = 640;
+			h = 480;
+		}
+		break;
 	case ImageQuality1M:
 		{
 			w = 1280;
diff --git a/src/storage/StorageManager.cpp b/src/storage/StorageManager.cpp
index e9f0449..0fa39a8 100755
--- a/src/storage/StorageManager.cpp
+++ b/src/storage/StorageManager.cpp
@@ -659,6 +659,63 @@ int StorageManager::generateFile(time_t &now, String8 &file, int cam_type, fileT
 	return fd;
 }
 
+int StorageManager::generateFile2(time_t now, String8 &file, int cam_type, fileType_t type)
+{
+
+    int fd;
+    struct tm *tm=NULL;
+    char buf[128]= {0};
+    char path_temp[128]= {0};
+    const char *camType = FILE_FRONT;
+    const char *suffix = FLAG_NONE;
+    const char *fileType = EXT_VIDEO_MP4;
+    const char *path = CON_2STRING(MOUNT_PATH, VIDEO_DIR);
+
+    file_threshold_t *threshlod_file = findThreshold(cam_type,type);
+    sprintf(path_temp,"%s%s/%s",MOUNT_PATH,threshlod_file->dir,THUMBNAIL_DIR);
+    path = path_temp;
+
+    if(isMount()== false)
+        return RET_NOT_MOUNT;
+    db_msg("%s %d", __FUNCTION__, __LINE__);
+
+    filelock.lock();
+    while(mFileMgrT == NULL)
+    {
+        mFileMgrT = new FileMgrThread(this);
+        status_t err = mFileMgrT->start();
+        if (err != OK) {
+            mFileMgrT.clear();
+            mFileMgrT = NULL;
+            db_msg("---FileMgrThread init error!!!!!");
+        }
+    }
+    thresholds.push_back(threshlod_file);
+    filelock.unlock();
+
+    tm = localtime(&now);
+
+    if (type == VIDEO_TYPE_IMPACT) {
+        suffix = FLAG_SOS;
+    }
+    if (type == PHOTO_TYPE_NORMAL) {
+        fileType = EXT_PIC_JPG;
+    }
+    if (cam_type == CAM_UVC) {
+        camType = FILE_BACK;
+    }
+    sprintf(buf, "%s%04d%02d%02d_%02d%02d%02d%s%s%s",path,tm->tm_year + 1900, tm->tm_mon + 1,
+            tm->tm_mday, tm->tm_hour,tm->tm_min, tm->tm_sec, camType, suffix, fileType);
+    file = buf;
+    fd = open(buf, O_RDWR | O_CREAT, 0666);
+    if (fd < 0) {
+        db_msg("failed to open file '%s'!!\n", buf);
+        return RET_IO_OPEN_FILE;
+    }
+    db_debug("generate file %s", file.string());
+    return fd;
+}
+
 int StorageManager::doMount(const char *fsPath, const char *mountPoint,
 		bool ro, bool remount, bool executable,
 		int ownerUid, int ownerGid, int permMask, bool createLost) {
diff --git a/src/window/RecordPreview.cpp b/src/window/RecordPreview.cpp
index be5de82..a705224 100755
--- a/src/window/RecordPreview.cpp
+++ b/src/window/RecordPreview.cpp
@@ -306,6 +306,7 @@ void RecordPreview::takePicFinish(int cam_id)
 {
 	Mutex::Autolock _l(mLockPic);
 	mAllowTakePic[cam_id] = true;
+	isThumb = false;
 }
 
 
@@ -357,7 +358,6 @@ static void jpegCallback(void *data, int size, void* caller, int id)
 		return;
 	}
 	rp = (RecordPreview*)caller;
-
 	rp->savePicture(data, size, id);
 	rp->takePicFinish(id);
 
@@ -684,6 +684,12 @@ void RecordPreview::recordListener(CdrMediaRecorder* cmr, int what, int extra)
 			}
 			db_msg("Please set the next fd.");
 			fd = sm->generateFile(mBakNow[cameraID], mBakFileString[cameraID], cameraID, recordParam.video_type);
+			isThumb = true;
+			#ifdef MESSAGE_FRAMEWORK
+			fwk_send_message_ext(FWK_MOD_VIEW,FWK_MOD_CONTROL, MSG_ID_FWK_CONTROL_TAKE_PICTURE_REQ, NULL, 0);
+			#else
+			takePicture();
+			#endif
 			if(fd <= 0) {
 				db_error("get fd failed\n");
 				clrDeleteFlag(cameraID, mRecordDBInfo[cameraID].fileString);
@@ -1017,6 +1023,7 @@ RecordPreview::RecordPreview(CdrMain *cdrMain)
 	mOldRPWindowState(STATUS_RECORDPREVIEW), 
 	mUvcPlugFlag(false),
 	SdcardState(true),
+	isThumb(false),
 #ifdef WATERMARK_ENABLE
 	mNeedWaterMark(false),
 #endif
@@ -1677,8 +1684,20 @@ int RecordPreview::prepareRecord(int cam_type, CdrRecordParam_t param)
 	int framerate = BACK_CAM_FRAMERATE;
 	int i = 0;
 	StorageManager* sm = StorageManager::getInstance();
-
+	
+	isThumb = true;
+	mBakFileString[cam_type].clear();
+	mBakNow[cam_type] = 0;
+	
 	fd = sm->generateFile(now, file, cam_type, param.video_type);
+	mBakNow[0] = now;
+	mBakFileString[0] = file;
+	#ifdef MESSAGE_FRAMEWORK
+	fwk_send_message_ext(FWK_MOD_VIEW,FWK_MOD_CONTROL, MSG_ID_FWK_CONTROL_TAKE_PICTURE_REQ, NULL, 0);
+	#else
+	takePicture();
+	#endif
+	
 	if (fd <= 0) {
 		SdcardState = false;
 		db_error("fail to generate file, retval is %d\n", fd);
@@ -1786,8 +1805,6 @@ int RecordPreview::startRecord(int cam_type, CdrRecordParam_t param)
 			SendMessage(hStatusBar, STBM_START_RECORD_TIMER, 0, 0);
 		}
 #endif
-		mBakFileString[cam_type].clear();
-		mBakNow[cam_type] = 0;
 		retval = mRecorder[cam_type]->start();
                 mRecorder[cam_type]->setSilent(mSilent);
 	}
@@ -2452,7 +2469,7 @@ void RecordPreview::takePicture()
 	StorageManager* sm;
 	ResourceManager* rm;
 	bool takePicture_status = false;
-
+	
 	if(mCamera[CAM_CSI] || mCamera[CAM_UVC]) {
 		sm = StorageManager::getInstance();
 		if(sm->isMount() == false) {
@@ -2470,13 +2487,23 @@ void RecordPreview::takePicture()
 			if (mCamera[CAM_CSI] && mCamera[CAM_CSI]->isPreviewing()) {
 				if (mAllowTakePic[CAM_CSI]) {
 					db_msg("csi take pic");
+					
+					if(isThumb == true)
+					{
+						mCamera[CAM_CSI]->setImageQuality(ImageQualityQQVGA);
+					}
+					else
+					{
+						mCamera[CAM_CSI]->setImageQuality(ImageQuality1M);
+					}
+				
 					mAllowTakePic[CAM_CSI] = false;
 #ifdef HAVE_takePicture
 					mCamera[CAM_CSI]->takePicture();
 					takePicture_status = true;
 #endif
 				} else {
-					db_msg("uvc last action hasn't finished");
+					db_msg("csi last action hasn't finished");
 				}
 			}
 			if (mCamera[CAM_UVC] && mCamera[CAM_UVC]->isPreviewing()) {
@@ -2527,8 +2554,16 @@ void RecordPreview::savePicture(void*data, int size, int id)
 
 	sm = StorageManager::getInstance();
 	hStatusBar = mCdrMain->getWindowHandle(WINDOWID_STATUSBAR);
-
-	fd = sm->generateFile(now, file, id, PHOTO_TYPE_NORMAL);
+	
+	if(isThumb == true)
+	{
+		fd = sm->generateFile2(mBakNow[0], file, 0, PHOTO_TYPE_NORMAL);
+	}
+	else
+	{
+		fd = sm->generateFile(now, file, id, PHOTO_TYPE_NORMAL);
+	}
+	
 	if(fd <= 0) {
 		db_error("generate picture file failed, retval is %d\n", fd);
 		return;	
